core/Pos/Core/Genesis/Testnet.hs

```diff
@@ -0,0 +1,217 @@
+                      (mkCoin $ fromIntegral faoOneBalance)
+
+    pure ([(gcdAddresses, gcdDistribution)], map snd fakeAvvmPubkeysAndSeeds)
+
+----------------------------------------------------------------------------
+-- Helpers
+----------------------------------------------------------------------------
+
+generateSecretsAndAddress
+    :: (MonadRandom m)
+    => Maybe (SecretKey, EncryptedSecretKey)  -- ^ plain key & hd wallet root key
+    -> m (SecretKey, EncryptedSecretKey, VssKeyPair, VssCertificate, Address)
+    -- ^ secret key, vss key pair, vss certificate,
+    -- hd wallet account address with bootstrap era distribution
+generateSecretsAndAddress mbSk = do
+    (sk, hdwSk, vss) <- generateSecrets mbSk
+
+    expiry <- fromInteger <$>
+        randomNumberInRange (Const.vssMinTTL - 1) (Const.vssMaxTTL - 1)
+    let vssPk = asBinary $ toVssPublicKey vss
+        vssCert = mkVssCertificate sk vssPk expiry
+        -- This address is used only to create genesis data. We don't
+        -- put it into a keyfile.
+        hdwAccountPk =
+            fst $ fromMaybe (error "generateKeyfile: pass mismatch") $
+            deriveLvl2KeyPair (IsBootstrapEraAddr True) emptyPassphrase hdwSk
+                Const.accountGenesisIndex Const.wAddressGenesisIndex
+    pure (sk, hdwSk, vss, vssCert, hdwAccountPk)
+
+generateFakeAvvm :: MonadRandom m => m (RedeemPublicKey, ByteString)
+generateFakeAvvm = do
+    seed <- getRandomBytes 32
+    let (pk, _) = fromMaybe
+            (error "Impossible - seed is not 32 bytes long") $
+            redeemDeterministicKeyGen seed
+    pure (pk, seed)
+
+generateSecrets
+    :: (MonadRandom m)
+    => Maybe (SecretKey, EncryptedSecretKey)
+    -> m (SecretKey, EncryptedSecretKey, VssKeyPair)
+generateSecrets mbSk = do
+    -- plain key & hd wallet root key
+    (sk, hdwSk) <-
+        case mbSk of
+            Just x -> return x
+            Nothing ->
+                (,) <$> (snd <$> keyGen) <*>
+                (snd <$> safeKeyGen emptyPassphrase)
+    vss <- vssKeyGen
+    pure (sk, hdwSk, vss)
+
+-- | Generates balance distribution for testnet.
+genTestnetDistribution :: TestnetBalanceOptions -> BalanceDistribution
+genTestnetDistribution TestnetBalanceOptions{..} =
+    checkConsistency $ RichPoorBalances {..}
+  where
+    richs = fromIntegral tboRichmen
+    poors = fromIntegral tboPoors * 2  -- for plain and hd wallet keys
+    testBalance = fromIntegral tboTotalBalance
+
+    -- Calculate actual balances
+    desiredRichBalance = getShare tboRichmenShare testBalance
+    oneRichmanBalance = desiredRichBalance `div` richs +
+        if desiredRichBalance `mod` richs > 0 then 1 else 0
+    realRichBalance = oneRichmanBalance * richs
+    poorsBalance = testBalance - realRichBalance
+    onePoorBalance = if poors == 0 then 0 else poorsBalance `div` poors
+    realPoorBalance = onePoorBalance * poors
+
+    mpcBalance = getShare (coinPortionToDouble Const.genesisMpcThd) testBalance
```

Comment:
`genesisMpcThd` isn't actually a constant, it's a configuration which depends upon a genesis which isn't known until run time (made clear in https://github.com/input-output-hk/cardano-sl/pull/1550).

So this is puzzling: in order to `genTestnetDistribution`, we need to have some genesis value. But presumably the testnet distribution is going to be used to define the genesis, right?

Reply:
i.e. there may be a cycle lurking here which would cause looping: `genTestnetDistribution` will hang on the `newInitFuture` `MVar` which it will attempt to fill after generating a genesis from that distribution. In #1550 this would be eliminated at compile time due to an unsatisfied constraint.

Reply:
@avieth this module generates `GenesisData` (actually part of it), while `genesisMpcThd` is known from `GenesisSpec`. So there is now loop here.
`GenesisSpec` is initialized by `applyConfigInfo`.

Reply:
I see. Good!
So do I understand correctly that there must always be a `GenesisData`, which is the single source of truth for the application, and there are precisely two ways of getting it:

```
-- Parse it from the canonical JSON
mainNetGenesisData :: FilePath -> IO (Either Error GenesisData)
-- Generate it from a spec
testNetGenesisData :: GenesisSpec -> Either Error GenesisData
```

?

Reply:
More or less yes.

For now we just want to support testnet way when we have `genesis-spec.yaml` (or, if we take your branch's approach, just some file on disk) with serialized `GenesisSpec`, parse it as soon as possible after application start and then use `generateTestnetOrMainnetData` to generate `GenesisData`. We are focused on making it work in all cases, it's CSLREQ-213.

After that we are planning two support two modes: passing either encoded `GenesisSpec` or encoded `GenesisData`. It's described in CSLREQ-214. There is some outdated info there, specifically, the fields from (1) shouldn't be optional, they are part of `GenesisSpec`. In `constants.yaml` we are going to have `genesis: genesis-mainnet.json` or `  genesisSpec: genesis-spec-testnet.yaml`.

It was our plan so far. I know that in your branch there is only one configuration file, so probably some adjustments are necessary.